src/consensus/tower.zig

```diff
@@ -1982,136 +1470,729 @@ pub const Tower = struct {
+/// common ancestor of both `slot_a` and `slot_b`.
+///
+/// Returns `null` if:
+/// * `slot_a` is not in `ancestors`
+/// * `slot_b` is not in `ancestors`
+/// * There is no common ancestor of slot_a and slot_b in `ancestors`
+fn greatestCommonAncestor(
+    ancestors: *const AutoHashMapUnmanaged(
+        Slot,
+        SortedSet(Slot),
+    ),
+    slot_a: Slot,
+    slot_b: Slot,
+) ?Slot {
+    var ancestors_a = ancestors.get(slot_a) orelse return null;
+    var ancestors_b = ancestors.get(slot_b) orelse return null;
 
-        try std.testing.expectError(
-            error.StakeNotFound,
-            tower.selectVoteAndResetForks(
-                std.testing.allocator,
-                100,
-                null,
-                100,
-                &.{},
-                &.{},
-                &ProgressMap.INIT,
-                &latest,
-                &fork_choice,
-                .{},
-                &slot_history,
-            ),
-        );
+    var max_slot: ?Slot = null;
+
+    var superset, const subset = if (ancestors_a.count() >= ancestors_b.count())
+        .{ ancestors_a, ancestors_b }
+    else
+        .{ ancestors_b, ancestors_a };
+
+    if (superset.count() == 0 or subset.count() == 0) return null;
+
+    for (superset.items()) |slot| {
+        if (!subset.contains(slot)) continue;
+        max_slot = if (max_slot) |current_max| @max(current_max, slot) else slot;
     }
 
-    /// Handles fork selection when switch fails due to duplicate rollback
-    pub fn selectCandidatesFailedSwitchDuplicateRollback(
-        allocator: std.mem.Allocator,
-        heaviest_slot: Slot,
-        // [Audit] Only used in logging in Agave
-        _: Slot,
-        failure_reasons: *std.ArrayListUnmanaged(HeaviestForkFailures),
-        initial_switch_fork_decision: SwitchForkDecision,
-    ) !CandidateVoteAndResetSlots {
-        // If we can't switch and our last vote was on an unconfirmed duplicate slot,
-        // we reset to the heaviest bank (even if not descendant of last vote)
+    return max_slot;
+}
 
-        try failure_reasons.append(allocator, .{
-            .FailedSwitchThreshold = .{
-                .slot = heaviest_slot,
-                .observed_stake = 0,
-                .total_stake = 0,
-            },
-        });
+// Optimistically skip the stake check if casting a vote would not increase
+// the lockout at this threshold. This is because if you bounce back to
+// voting on the main fork after not voting for a while, your latest vote
+// might pop off a lot of the votes in the tower. The stake from these votes
+// would have rolled up to earlier votes in the tower, which presumably
+// could have helped us pass the threshold check. Worst case, we'll just
+// recheck later without having increased lockouts.
+fn optimisticallyBypassVoteStakeThresholdCheck(
+    // Needs to be an iterator that produces Lockout
+    tower_before_applying_vote: anytype,
```

Comment:
This comment appears to be inaccurate/misleading, and this type could maybe be changed to just be a `[]const Slot`?